### PR TITLE
Speed up reply<->state check

### LINKED ISSUE #161: Speed up the config diff
The config diff takes forever (~1s per config being checked). However there's usually only a few variations at desync. Hashing a serialised config is super quick (~15ms), so we should probably build up a cache of hashes to `DeepDiff`s as we go through it. You'd simply use a `dict` which maps a tuple of `(hash(state_value), hash(host_value)) -> DeepDiff`

Doing this I imagine we'd be able to get the desync check down from 3m30s to more like 30s, which is way better.

### PR DESCRIPTION
**Pull request by nritsche**
Speeds up the kotekan config check by minimizing the number of DeepDiff calls.

There's a bit of refactoring and test additions to facilitate the change, but the commit messages hopefully make the review easy.

The two `fix` commits fix an issue discovered when adding the test.

Closes #161

### CODE REVIEW

File: coco/check.py [hunk 1]
```diff
@@ -377,23 +377,15 @@ async def run(self, result: Result):
...
-                if self.state_paths:
+        if self.state_paths:
+            for host, result_ in reply.items():
+                if not result_:
                     for name in self.state_paths.keys():
                         logger.debug(
                             f"/{self._name}: Missing value '{name}' in reply from {host}."
                         )
                         failed_hosts.add(host)
                         result.report_failure(self._name, host, "missing", name)
```
Comment (tristpinsm) [inline coco/check.py:393]: there used to be a `continue` at the end of this block. Won't `result_.items()` fail `if not result_`?

File: coco/check.py [hunk 2]
```diff
@@ -342,11 +342,12 @@ def __init__(self, name, state_paths, on_failure, save_to_state, forwarder, stat
                         f"Found value '{field}' of type '{type(path).__name__}' "
                         f"in state-reply-check for /{name} (expected 'str')."
                     )
-                if not state.find_or_create(path):
+                if not state.exists(path):
```
Comment (tristpinsm) [inline coco/check.py:345]: isn't the other case where `state_paths` is a `str` (line 331) also affected by this issue and should be changed too?

File: coco/check.py [hunk 3]
```diff
@@ -377,23 +377,15 @@ async def run(self, result: Result):
             if r:
                 reply.update(r)
 
-        for host, result_ in reply.items():
-            if not result_:
-                if self.state_paths:
+        if self.state_paths:
```
Comment (anjakefala) [inline coco/check.py:385]: What is the difference between self.state_paths and self.state_path?

**Comment by anjakefala** [comment]
I did not notice anything outside of the things Tristan has already commented on. =)

File: coco/check.py (hunk 3 above)
Comment (nritsche) [reply coco/check.py:385]: I think it's just an ugly way to keep either a single path or multiple.

File: coco/check.py (hunk 1 above)
Comment (nritsche) [reply coco/check.py:393]: Thanks, good catch!

File: coco/check.py (hunk 2 above)
Comment (nritsche) [reply coco/check.py:346]: Yeah, you're right!

**Comment by nritsche** [comment]
Thanks for reviewing!

**Pull request merged**